Dockerfile.e2e-tests

```diff
@@ -0,0 +1,18 @@
+# Taking a non-alpine image for e2e tests so that cgo can be enabled for the race detector.
+FROM golang:1.21 as builder
+
+WORKDIR $GOPATH/src/github.com/thanos-io/thanos
+
+COPY . $GOPATH/src/github.com/thanos-io/thanos
+
+RUN CGO_ENABLED=1 go build -o $GOBIN/thanos -race ./cmd/thanos
+# -----------------------------------------------------------------------------
+
+FROM golang:1.21
+LABEL maintainer="The Thanos Authors"
+
+COPY --from=builder $GOBIN/thanos /bin/thanos
+
+ENV GORACE="halt_on_error=1"
```

Comment:
@GiedriusS is this what we really want? We have plenty of e2e tests and they take very long to run. They will take even more time with race detection enabled. 

I think it would be useful to see a full report of the test suite even when something fails at the very beginning so that someone working on the codebase can quickly take notice of all the races detected instead of seeing one, fixing, waiting for the build to see another one, fixing, etc. What do you think?

Reply:
In the past we explicitly removed the "half on error" behavior of the e2e tests exactly for this same reason.

Reply:
Ah no, my bad. We've had fail-fast since forever in the e2e tests. I think my brain has some crossed wires, sorry. 😂

Reply:
I ran the e2e tests with race detector many times so I'm quite confident that people shouldn't run into random failures. They are a bit slower but not by much. But even without the race detector, the tests sometimes run on very CPU constrained nodes so maybe it makes overall sense in the near future to use dedicated workers paid for by CNCF